frappe/core/doctype/user/user.py

```diff
@@ -167,23 +171,47 @@ def validate(self):
 		self.validate_allowed_modules()
 		self.validate_user_image()
 		self.set_time_zone()
-
 		if self.language == "Loading...":
 			self.language = None
 
 		if (self.name not in ["Administrator", "Guest"]) and (not self.get_social_login_userid("frappe")):
 			self.set_social_login_userid("frappe", frappe.generate_hash(length=39))
 
 	def populate_role_profile_roles(self):
-		if self.role_profile_name:
-			role_profile = frappe.get_doc("Role Profile", self.role_profile_name)
+		roles = []
+		if self.role_profiles:
+			for role_profile in self.role_profiles:
+				role_profile = frappe.get_doc("Role Profile", role_profile.role_profile)
+				roles.extend([role.role for role in role_profile.roles])
 			self.set("roles", [])
-			self.append_roles(*[role.role for role in role_profile.roles])
+			self.append_roles(*roles)
 
 	@deprecated
 	def validate_roles(self):
 		self.populate_role_profile_roles()
 
+	def move_role_profile_name_to_role_profiles(self):
+		if not self.role_profile_name or any(
+			(profile.role_profile == self.role_profile_name) for profile in self.role_profiles
+		):
+			return
+		profile_exists = frappe.db.exists(
+			"User Role Profile",
+			{"parent": self.name, "role_profile": self.role_profile_name, "parenttype": "User"},
+		)
+		if not profile_exists:
+			user_role_profile = frappe.get_doc(
+				{
+					"doctype": "User Role Profile",
+					"role_profile": self.role_profile_name,
+					"parent": self.name,
+					"parenttype": "User",
+					"parentfield": "role_profiles",
+				}
+			)
+			user_role_profile.save(ignore_permissions=True)
+			self.role_profiles.append(user_role_profile)
```

Comment:
This is needlessly complicated, you have access to child table, just verify it from there directly.